tokio/src/task/local.rs

```diff
@@ -0,0 +1,585 @@
+//! Runs `!Send` futures on the current thread.
+use crate::task::{self, JoinHandle, Schedule, UnsendTask, Unsendable};
+
+use std::cell::{Cell, UnsafeCell};
+use std::collections::VecDeque;
+use std::fmt;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::ptr::{self, NonNull};
+use std::rc::Rc;
+use std::task::{Context, Poll};
+
+use pin_project::pin_project;
+
+/// A set of tasks which are executed on the same thread.
+///
+/// In some cases, it is necessary to run one or more futures that do not
+/// implement [`Send`] and thus are unsafe to send between threads. In these
+/// cases, a [local task set] may be used to schedule one or more `!Send`
+/// futures to run together on the same thread.
+///
+/// For example, the following code will not compile:
+///
+/// ```rust,compile_fail
+/// # use tokio::runtime::Runtime;
+/// use std::rc::Rc;
+///
+/// // `Rc` does not implement `Send`, and thus may not be sent between
+/// // threads safely.
+/// let unsend_data = Rc::new("my unsend data...");
+///
+/// let mut rt = Runtime::new().unwrap();
+///
+/// rt.block_on(async move {
+///     let unsend_data = unsend_data.clone();
+///     // Because the `async` block here moves `unsend_data`, the future is `!Send`.
+///     // Since `tokio::spawn` requires the spawned future to implement `Send`, this
+///     // will not compile.
+///     tokio::spawn(async move {
+///         println!("{}", unsend_data);
+///         // ...
+///     }).await.unwrap();
+/// });
+/// ```
+/// In order to spawn `!Send` futures, we can use a local task set to
+/// schedule them on the thread calling [`Runtime::block_on`]. When running
+/// inside of the local task set, we can use [`task::spawn_local`], which can
+/// spawn `!Send` futures. For example:
+///
+/// ```rust
+/// # use tokio::runtime::Runtime;
+/// use std::rc::Rc;
+/// use tokio::task;
+///
+/// let unsend_data = Rc::new("my unsend data...");
+///
+/// let mut rt = Runtime::new().unwrap();
+/// // Construct a local task set that can run `!Send` futures.
+/// let local = task::LocalSet::new();
+///
+/// // Run the local task group.
+/// local.block_on(&mut rt, async move {
+///     let unsend_data = unsend_data.clone();
+///     // `spawn_local` ensures that the future is spawned on the local
+///     // task group.
+///     task::spawn_local(async move {
+///         println!("{}", unsend_data);
+///         // ...
+///     }).await.unwrap();
+/// });
+/// ```
+///
+/// [`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
+/// [local task set]: struct.LocalSet.html
+/// [`Runtime::block_on`]: ../struct.Runtime.html#method.block_on
+/// [`task::spawn_local`]: fn.spawn.html
+#[derive(Debug)]
+pub struct LocalSet {
+    scheduler: Rc<Scheduler>,
+    _not_send_or_sync: PhantomData<*const ()>,
```

Comment:
Is this `PhantomData` required w/ the `Rc`? I would add a comment explaining why it is needed.

Reply:
ah, I think that's an artifact of before the Rc was added, I missed that. I think it can safely be removed!

Reply:
ah, I think that's an artifact of before the Rc was added, I missed that. I think it can safely be removed!